### ISSUE TITLE
Terraform modules throws random unclear errors.

### ISSUE THREAD

**Issue by HebertCL** [hashicorp/terraform#14324]
Hi,

I have been trying to use Terraform modules in combination with tfvars/vars files to create several instances, auto scaling groups, launch configurations and security groups in AWS, but every time I get random unclear errors. I have tested several posible options that I will sumarize here:

 ### Module Structure

I have three scripts using modules, all of them with the following structure:

```hcl
.
├── jump_preprod
│   ├── instance.tf
│   ├── remote_config.tf
│   ├── security.tf
│   └── vars.tf
├── jump_prod
│   ├── instance.tf
│   ├── remote_config.tf
│   ├── security.tf
│   └── vars.tf
├── main.tf
├── terraform.tfvars
└── vars.tf
```

I can run `terraform get` and `terraform validate` with no errors. Once I run `terraform plan` or `terraform apply` I get the following error:

```hcl
terraform plan
1 error(s) occurred:

* module root:
    module preprod: key_name is not a valid parameter
    module preprod: volume_type is not a valid parameter
    module preprod: volume_size is not a valid parameter
    module preprod: ami is not a valid parameter
    module preprod: instance_type is not a valid parameter
    module preprod: required variable "jumppp_instance" not set
    module preprod: required variable "logging_key" not set
    module preprod: required variable "jumppp_ebs_type" not set
    module preprod: required variable "jumppp_ebs_size" not set
    module preprod: required variable "jumppp_ami" not set
    module prod: instance_type is not a valid parameter
    module prod: key_name is not a valid parameter
    module prod: volume_type is not a valid parameter
    module prod: volume_size is not a valid parameter
    module prod: ami is not a valid parameter
    module prod: required variable "jumprod_instance" not set
    module prod: required variable "logging_key" not set
    module prod: required variable "jumprod_ebs_type" not set
    module prod: required variable "jumprod_ebs_size" not set
    module prod: required variable "jumprod_ami" not set
```

The result is the same running Terraform v0.9.4 and v0.9.3. Terraform will throw error stating each parameter used in the modules is not valid and not set. I decided to switch back to v0.8 and test to see if it was an issue related to version 0.9. I have tested with v0.8.7 and v0.8.6, both with the same following result:

```hcl
terraform plan
module root: module preprod: ami is not a valid parameter
terraform plan
module root: module preprod: instance_type is not a valid parameter
 terraform plan
module root: module preprod: volume_size is not a valid parameter
```
Each time I throw `terraform plan` a random error comes out. The same issue was previously reported in **issue #1652**. I was able of sort this out previously eliminating variables until Terraform stopped throwing errors. I am not sure how my workaround removing variables actually worked, but it is not practical in terms of script reusability.

### Using tfvars in each module

I have tested with `terraform.tfvars` file in each module to make sure variables match and are not causing this errors. When I run `terraform apply` with tfvars file inside the modules:

```hcl
terraform plan
+ aws_instance.jump_preprod
    ami:                                               "ami-4191b524"
    associate_public_ip_address:                       "false"
    availability_zone:                                 "<computed>"
...
+ aws_security_group.jumpppsg
    description:                          " Launch Configuration Security Group"
    egress.#:                             "1"
    egress.482069346.cidr_blocks.#:       "1"
    egress.482069346.cidr_blocks.0:       "0.0.0.0/0"

...
Plan: 2 to add, 0 to change, 0 to destroy.
```

This behavior is the same across all my modules, with Terraform 0.9.x and 0.8.x. I did testing on several machines with versions 0.8.8 and 0.9.4 with same results I described above. This was also tested in Ubuntu 16.04, CentOS 7 (not sure if this makes any difference since both use the same binary) and Windows 10 with the same results. I did some research and only found **issue #12479** with some relation with my own issues. I appreciate any feedback or solution.

Thanks in advance.

### Terraform Version
v0.8.6, v0.8.7, v0.9.3, v0.9.4

### Affected Resource(s)
Multiple resources, might be a core issue:
- aws_instance
- aws_launch_configuration
- aws_autoscaling_group
- aws_security_group

### Terraform Configuration Files

Too many to paste here. This is the link to my repository, you can clone and test with it:
https://github.com/HebertCL/Learning_Terraform/tree/master/functional_environment

### Expected Behavior

Display the plan/Apply for the resource creation.

### Actual Behavior

Terraform throws either a random error or a bunch of errors related to parameters not recognized and not set. 

### Steps to Reproduce

1. `terraform get`
2. `terraform validate`
3. `terraform plan`

### References

- GH-1652
- GH-12479

**Comment by HebertCL** [comment]
### Update:

I have downloaded and tested with v0.9.5. The result is the same as with v0.9.4 and v0.9.3.

**Comment by HebertCL** [comment]
# Update:

I have been running some tests with latest version (0.9.5) and 0.8.6 without using terraform.tfvars file, entering all variables by hand for testing purposes. These are the results:

### v0.9.5:

Using latest Terraform version I am able to enter all values by hand, but will fail in the same way described before.

```hcl
hebert@heberttf-VB:~/Learning_Terraform/functional_environment/management/jump_hosts$ terraform plan
var.jumppp_ami
  Enter a value: ami-4191b524

var.jumppp_ebs_size
  Enter a value: 2

var.jumppp_ebs_type
  Enter a value: standard

var.jumppp_instance
  Enter a value: t2.micro

var.logging_key
  Enter a value: hacl

1 error(s) occurred:

* module root:
    module preprod: volume_size is not a valid parameter
    module preprod: ami is not a valid parameter
    module preprod: instance_type is not a valid parameter
    module preprod: key_name is not a valid parameter
    module preprod: volume_type is not a valid parameter
    module preprod: required variable "jumppp_ebs_type" not set
    module preprod: required variable "jumppp_ebs_size" not set
    module preprod: required variable "jumppp_ami" not set
    module preprod: required variable "jumppp_instance" not set
    module preprod: required variable "logging_key" not set
```

### v0.8.6:

Using this version of Terraform will throw an error getting the module and planning. This is the random behavior described previously with versions 0.8.x.

```hcl
hebert@heberttf-VB:~/Learning_Terraform/functional_environment/management/jump_hosts$ terraform v      ersion
Terraform v0.8.6

Your version of Terraform is out of date! The latest version
is 0.9.5. You can update by downloading from www.terraform.io
hebert@heberttf-VB:~/Learning_Terraform/functional_environment/management/jump_hosts$ terraform get
Get: file:///home/hebert/Learning_Terraform/functional_environment/management/jump_hosts/jump_preprod
Error loading Terraform: module root: module preprod: instance_type is not a valid parameter
hebert@heberttf-VB:~/Learning_Terraform/functional_environment/management/jump_hosts$ terraform plan
module root: module preprod: instance_type is not a valid parameter
hebert@heberttf-VB:~/Learning_Terraform/functional_environment/management/jump_hosts$
```

Running this test I was expecting it to work and conclude that somehow Terraform is ignoring tfvars file. This is clearly not the case and Terraform will also fail if values are entered manually. For v0.8.x behavior remains the same.

**Comment by johnlabarge** [comment]
I'm having something similar.  First experience using terraform so I'm a noob, but I don't fully understand the error message.

**Comment by HebertCL** [comment]
@johnlabarge might be helpful if you post your scripts and the error you're facing.

**Comment by josephholsten** [comment]
@herbertcl want to pick one of these error messages for me to start trying to write a fix for? What do we want the error message to be?

**Comment by HebertCL** [comment]
@josephholsten correct, I want to try a workaround for this if possible. Just let me know if you get the same errors I get with the same resources I used, or if any other resources you used giving the same behavior.

**Comment by mMeijden** [comment]
I've experienced similar problems in 0.9.11. It occurred in my test environment so i tried it with a terraform destroy and that solved all the problems... I can imagine you don't want to that in production :/

**Comment by marco-m** [comment]
hello any news ?

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

There were quite a few different issues covering this issue of un-actionable error messages, and so we didn't do a good job of keeping up to date on them all, but I'm pleased to report that better error messages are now in master and ready to be included in the forthcoming v0.12.0 release. You can see some examples over in #7189.

Thanks for reporting this!
